sympy/ntheory/residue_ntheory.py

```diff
@@ -940,46 +998,39 @@ def nthroot_mod(a, n, p, all_roots=False):
     [8, 11]
     >>> nthroot_mod(68, 3, 109)
     23
+
+    References
+    ==========
+
+    .. [1] P. Hackman "Elementary Number Theory" (2009), page 76
+
     """
     a = a % p
     a, n, p = as_int(a), as_int(n), as_int(p)
 
+    if n < 1:
+        raise ValueError("n should be positive")
+    if p < 1:
+        raise ValueError("p should be positive")
+    if n == 1:
+        return [a] if all_roots else a
     if n == 2:
         return sqrt_mod(a, p, all_roots)
-    # see Hackman "Elementary Number Theory" (2009), page 76
-    if not isprime(p):
-        return _nthroot_mod_composite(a, n, p)
-    if a % p == 0:
-        return [0]
-    if not is_nthpow_residue(a, n, p):
-        return [] if all_roots else None
-    if (p - 1) % n == 0:
-        return _nthroot_mod1(a, n, p, all_roots)
-    # The roots of ``x**n - a = 0 (mod p)`` are roots of
-    # ``gcd(x**n - a, x**(p - 1) - 1) = 0 (mod p)``
-    pa = n
-    pb = p - 1
-    b = 1
-    if pa < pb:
-        a, pa, b, pb = b, pb, a, pa
-    while pb:
-        # x**pa - a = 0; x**pb - b = 0
-        # x**pa - a = x**(q*pb + r) - a = (x**pb)**q * x**r - a =
-        #             b**q * x**r - a; x**r - c = 0; c = b**-q * a mod p
-        q, r = divmod(pa, pb)
-        c = pow(b, -q, p) * a % p
-        pa, pb = pb, r
-        a, b = b, c
-    if pa == 1:
-        if all_roots:
-            res = [a]
-        else:
-            res = a
-    elif pa == 2:
-        return sqrt_mod(a, p, all_roots)
-    else:
-        res = _nthroot_mod1(a, pa, p, all_roots)
-    return res
+    base = []
```

Comment:
This is a new algorithm, right? I see you modified the testing. Do you know how the performance compares?

Reply:
> This is a new algorithm, right? I see you modified the testing.

Yes. Always return int or None if `all_roots=False`.

> Do you know how the performance compares?

I'm not sure. If you don't mind, could you please let us know?

Reply:
> Yes. Always return int or None

It's significantly different than that. The former routine started out with a few quick-exit tests. Where is the routine of this PR from?

Reply:
This is the processing that was done by a part of `_help`.
Previously, `_nthroot_mod_composite` (`_help`) or `nthroot_mod` did the processing, depending on whether the number was composite or prime, respectively. However, in the case of composite numbers, `nthroot_mod` is called further internally, which is a little confusing.
So, I rearranged it so that `nthroot_mod` factorizes `p` into prime factors and calls `_nthroot_mod_prime_power`.

Reply:
I made a diagram of the image.

![20230621](https://github.com/sympy/sympy/assets/36563693/c908d201-50d0-45b2-8f26-b63e299ddc8b)

Reply:
Thanks for that. This is a very effective representation!

Reply:
Sorry for the lack of explanation.